### PR TITLE
KDESKTOP-953-Use-kStore-to-check-for-new-updates2

### PR DESCRIPTION
**Pull request by ClementKunz**

### CODE REVIEW

File: infomaniak/kDrive.cmake [hunk 1]
```diff
@@ -35,12 +35,14 @@ if("$ENV{KDRIVE_DEBUG}" STREQUAL "1")
     set( KDRIVE_API_V2_URL "https://api.kdrive.preprod.dev.infomaniak.ch/2" )
     set( NOTIFY_KDRIVE_V2_URL "https://notify.kdrive.preprod.dev.infomaniak.ch/2" )
     set( LOGIN_URL "https://login.preprod.dev.infomaniak.ch" )
+    set( INFOMANIAK_API_URL "https://api.infomaniak.com/1" )
```
Comment (luc-guyot-infomaniak) [inline infomaniak/kDrive.cmake:38]: If the two `set` statements are really the same, you can put this common statement outside of the `if-else` blocks.

File: src/3rdparty/keychain/test/catch.hpp
Comment (luc-guyot-infomaniak) [inline src/3rdparty/keychain/test/catch.hpp:1]: It looks as if this file should not be modified. Was it modified in another PR?

File: src/libcommon/utility/types.h [hunk 2]
```diff
@@ -23,11 +23,21 @@
 #include <functional>
 #include <cctype>
 #include <optional>
+#include <sstream>
 #include <unordered_set>
 #include <variant>
 
 namespace KDC {
 
+typedef int64_t SyncTime;
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/types.h:32]: Accidental additions of `typedef`s.

File: src/libcommon/utility/types.h [hunk 3]
```diff
@@ -332,19 +343,41 @@ enum class AppStateKey {
...
 enum class LogUploadState { None, Archiving, Uploading, Success, Failed, CancelRequested, Canceled };
-
 enum class UpdateState { Error, None, Checking, Downloading, Ready, ManualOnly, Skipped };
 
+enum class UpdateStateV2 { UpToDate, Available, Downloading, Ready, Error };
+enum class DistributionChannel { Prod, Next, Beta, Internal, Unknown };
+enum class Platform { MacOS, Windows, LinuxAMD, LinuxARM, Unknown };
+
+struct VersionInfo {
+        std::string tag;
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/types.h:361]: The `tag` member variable should be documented in. a docstring: Is it a `git` tag?

File: src/libcommon/utility/types.h [hunk 4]
```diff
@@ -332,19 +343,41 @@ enum class AppStateKey {
...
+        std::string changeLog;
+        std::uint64_t buildVersion = 0;
+        std::uint64_t buildMinOsVersion = 0;
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/types.h:364]: A docstring would be nice to have.

File: src/libcommon/utility/utility.cpp [hunk 5]
```diff
@@ -215,15 +238,15 @@ qint64 CommonUtility::freeDiskSpace(const QString &path) {
     return -1;
 }
 
-QByteArray CommonUtility::IntToArray(qint32 source) {
+QByteArray CommonUtility::intToArray(qint32 source) {
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/utility.cpp:241]:
```suggestion
QByteArray CommonUtility::toQByteArray(qint32 source) {
```

File: src/libcommon/utility/utility.cpp [hunk 6]
```diff
@@ -215,15 +238,15 @@ qint64 CommonUtility::freeDiskSpace(const QString &path) {
...
     // Avoid use of cast, this is the Qt way to serialize objects
     QByteArray temp;
     QDataStream data(&temp, QIODevice::ReadWrite);
     data << source;
     return temp;
 }
 
-int CommonUtility::ArrayToInt(QByteArray source) {
+int CommonUtility::arrayToInt(QByteArray source) {
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/utility.cpp:249]:
```suggestion
int CommonUtility::toInt(QByteArray source) {
```

**Review by luc-guyot-infomaniak — COMMENTED** [review]
A first pass.

File: src/3rdparty/keychain/test/catch.hpp
Comment (ClementKunz) [reply src/3rdparty/keychain/test/catch.hpp:1]: Very strange indeed. Looks like it was introduced in commit : 04237027da46550a3c90fa46063d65bd7e2e9d0e
I try to revert the change but it was even worse

File: src/libcommon/utility/types.h [hunk 7]
```diff
@@ -332,19 +334,42 @@ enum class AppStateKey {
...
 enum class LogUploadState { None, Archiving, Uploading, Success, Failed, CancelRequested, Canceled };
-
 enum class UpdateState { Error, None, Checking, Downloading, Ready, ManualOnly, Skipped };
 
+enum class UpdateStateV2 { UpToDate, Available, Downloading, Ready, Error };
+enum class DistributionChannel { Prod, Next, Beta, Internal, Unknown };
+enum class Platform { MacOS, Windows, LinuxAMD, LinuxARM, Unknown };
+
+struct VersionInfo {
+        std::string tag;                      // Version number. ex: 3.6.4
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/types.h:352]:
```suggestion
        std::string tag;                      // Version number, for example: 3.6.4
```

File: src/libsyncengine/jobs/network/abstractnetworkjob.cpp [hunk 8]
```diff
@@ -573,26 +584,118 @@ bool AbstractNetworkJob::processSocketError(const std::string &msg, const Unique
...
+    }
+
+    std::stringstream ss;
+    if (encoding == "gzip") {
+        unzip(is, ss);
+    } else {
+        ss << is.rdbuf();
+    }
+
+    Poco::JSON::Object::Ptr jsonObj;
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/jobs/network/abstractnetworkjob.cpp:670]: Most of the code below looks very similar to what we see in `AbstractNetworkJob::handleJsonResponse` implementation.

File: src/libsyncengine/jobs/network/getappversionjob.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    ss << "/app-information/kstore-update/" << platformToStr(_platform) << "/com.infomaniak.drive/" << _appId;
+    return ss.str();
+}
+std::string GetAppVersionJob::getContentType(bool &canceled) {
+    canceled = false;
+    return {};
+}
+
+bool GetAppVersionJob::handleError(std::istream &is, const Poco::URI &uri) {
+    // TODO
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/jobs/network/getappversionjob.cpp:55]: Is it planned for this PR or for another one?

File: src/libsyncengine/jobs/network/networkjobsparams.h [hunk 10]
```diff
@@ -67,6 +67,21 @@ static const std::string cursorKey = "cursor";
 static const std::string hasMoreKey = "has_more";
 static const std::string tokenKey = "token";
 static const std::string preferenceKey = "preference";
+static const std::string applicationKey = "application";
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/jobs/network/networkjobsparams.h:70]: I am wondering whether these definitions of keys, or even just their litteral values, should into the body of `GetAppVersionJob::toDistributionChannel`?

File: src/libsyncengine/jobs/network/getappversionjob.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        std::string versionType;
+        if (!JsonParserUtility::extractValue(obj, typeKey, versionType)) {
+            return false;
+        }
+
+        const DistributionChannel channel = toDistributionChannel(versionType);
+        if (!JsonParserUtility::extractValue(obj, tagKey, _versionInfo[channel].tag)) {
+            return false;
+        }
+        if (!JsonParserUtility::extractValue(obj, changeLogKey, _versionInfo[channel].changeLog)) {
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/jobs/network/getappversionjob.cpp:100]: You could iterate on a vector of key strings.

File: src/libcommon/utility/types.h (hunk 7 above)
Comment (luc-guyot-infomaniak) [reply src/libcommon/utility/types.h:352]: Does the name `tag` reflects the fact that it coincides with the `git tag` of the release branch.

File: test/server/updater/testabstractupdater.cpp [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+void TestAbstractUpdater::setUp() {
+    LOG_DEBUG(Log::instance()->getLogger(), "$$$$$ Set Up");
+
+    const std::string userIdStr = loadEnvVariable("KDRIVE_TEST_CI_USER_ID");
```
Comment (luc-guyot-infomaniak) [inline test/server/updater/testabstractupdater.cpp:43]: It looks as it's time to make a re-usable function for this repeated task.

Comment (ClementKunz) [reply test/server/updater/testabstractupdater.cpp]: Done in commit a45b68e20e8c1f2b864f5d912e07f868a9987174

File: src/libsyncengine/jobs/network/getappversionjob.cpp (hunk 11 above)
Comment (ClementKunz) [reply src/libsyncengine/jobs/network/getappversionjob.cpp:117]: Not sure how. This is the JSON object I receive in the `published_versions`:
`
{
          "tag": "3.6.4",
          "tag_updated_at": "2024-08-14 12:00:00",
          "version_changelog": "Release notes - ...",
          "type": "production",
          "build_version": "20240814",
          "build_min_os_version": "",
          "download_link": "url",
          "data": [
            "[]"
          ]
}
`

I first need to extract the type. I do not see clearly how iterating over the item of the JSON object would simplify the code. Do you have something precise in mind?

File: src/libsyncengine/jobs/network/getappversionjob.cpp (hunk 9 above)
Comment (ClementKunz) [reply src/libsyncengine/jobs/network/getappversionjob.cpp:72]: In the next one, sorry. This one was becoming too big due to refactoring of the network job classes.

File: src/libsyncengine/jobs/network/abstractnetworkjob.cpp (hunk 8 above)
Comment (ClementKunz) [reply src/libsyncengine/jobs/network/abstractnetworkjob.cpp]: Done

File: src/libsyncengine/jobs/network/abstractnetworkjob.cpp [hunk 13]
```diff
@@ -573,26 +584,118 @@ bool AbstractNetworkJob::processSocketError(const std::string &msg, const Unique
...
+    // Extract reply
+    std::ostringstream os;
+    _jsonRes->stringify(os);
+    if (isExtendedLog()) {
+        LOGW_DEBUG(_logger, L"Reply " << jobId() << L" received: " << Utility::s2ws(os.str()).c_str());
+    }
+    return true;
+}
+
+bool AbstractNetworkJob::handleJsonResponse(std::string &str) {
```
[the remark this answers is not in the document]
Comment (ClementKunz) [reply src/libsyncengine/jobs/network/abstractnetworkjob.cpp]: Done

File: src/libsyncengine/jobs/network/API_v2/abstracttokennetworkjob.cpp [hunk 14]
```diff
@@ -18,11 +18,11 @@
 
 #include "abstracttokennetworkjob.h"
 #include "config.h"
-#include "jobs/network/networkjobsparams.h"
-#include "libcommonserver/utility/utility.h"
-#include "libparms/db/parmsdb.h"
-#include "keychainmanager/keychainmanager.h"
-#include "utility/jsonparserutility.h"
+#include "../networkjobsparams.h"
+#include "../../../../libcommonserver/utility/utility.h"
```
[the remark this answers is not in the document]
Comment (ClementKunz) [reply src/libsyncengine/jobs/network/API_v2/abstracttokennetworkjob.cpp]: Yes sorry. The refactoring of CLion added those, and I did not check

File: src/libcommon/utility/types.h [hunk 15]
```diff
@@ -332,19 +334,42 @@ enum class AppStateKey {
...
+        std::string changeLog;                // List of changes in this version
+        std::uint64_t buildVersion = 0;       // ex: 20240816
+        std::uint64_t buildMinOsVersion = 0;  // Minimum version of the OS supported. ex: 10.15, 11, ...
```
[the remark this answers is not in the document]
Comment (ClementKunz) [reply src/libcommon/utility/types.h:355]: You are right, it should not be an integral value.

File: src/libcommon/utility/types.h (hunk 7 above)
Comment (ClementKunz) [reply src/libcommon/utility/types.h:352]: As a matter of fact, it does. But no, it reflects the key retrieved from the GetAppVersion request,

File: test/server/updater/testabstractupdater.cpp [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+    private:
+        bool _updateAvailable{false};
+};
+
+void TestAbstractUpdater::testCheckUpdateAvailable() {
+    const auto appUid = "1234567890";
+
+    // Version is higher than current version
+    auto *testJob = new TestGetAppVersionJob(CommonUtility::platform(), appUid, true);
```
Comment (luc-guyot-infomaniak) [inline test/server/updater/testabstractupdater.cpp:72]: You can isolate each subtest with braces (like ´{ subtest }´) to make sure any of your temporary variables is misused by a subsequent test.

File: test/server/updater/testabstractupdater.cpp [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    AbstractUpdater::instance()->checkUpdateAvailable(updateAvailable);
+    CPPUNIT_ASSERT(updateAvailable);
+
+    // Version is lower than current version
+    testJob = new TestGetAppVersionJob(CommonUtility::platform(), appUid, false);
+    AbstractUpdater::instance()->setGetAppVersionJob(testJob);
+    AbstractUpdater::instance()->checkUpdateAvailable(updateAvailable);
+    CPPUNIT_ASSERT(!updateAvailable);
+
+    delete testJob;
```
Comment (luc-guyot-infomaniak) [inline test/server/updater/testabstractupdater.cpp:84]: You forgot to delete the previous test object.

File: test/server/updater/testabstractupdater.cpp [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+void TestAbstractUpdater::testCurrentVersion() {
+    std::string test = CommonUtility::currentVersion();
```
Comment (luc-guyot-infomaniak) [inline test/server/updater/testabstractupdater.cpp:88]:
```suggestion
    const std::string test = CommonUtility::currentVersion();
```

**Review by luc-guyot-infomaniak — COMMENTED** [review]
Minor comments. 

Side remark: Wouldn't it be better to store the newly introduced key strings in a class, or at least, in a namespace?

File: test/libsyncengine/jobs/network/testnetworkjobs.h [hunk 19]
```diff
@@ -48,13 +47,14 @@ class TestNetworkJobs : public CppUnit::TestFixture {
         CPPUNIT_TEST(testThumbnail);
         CPPUNIT_TEST(testDuplicateRenameMove);
         CPPUNIT_TEST(testRename);
-        CPPUNIT_TEST(testUpload);
+        // CPPUNIT_TEST(testUpload);
         //        CPPUNIT_TEST(testUploadAborted);
-        CPPUNIT_TEST(testDriveUploadSessionConstructorException);
```
[the remark this answers is not in the document]
Comment (ClementKunz) [reply test/libsyncengine/jobs/network/testnetworkjobs.h:53]: Indeed, I did, sorry

**Review by luc-guyot-infomaniak — APPROVED** [review]

**Pull request merged**